Logging Update

awholegunch opened the pull request:
This is a work-in-progress, as the diagrams need to be corrected and revised.

Quick Build: http://builds.cask.co/browse/CDAP-DQB303-10

Page of interest: **Logging and Monitoring:** http://builds.cask.co/artifact/CDAP-DQB303/shared/build-10/Docs-HTML/4.1.0/en/admin-manual/operations/logging.html

Commit:
w-i-p

Commit:
w-i-p

Commit:
w-i-p

Commit:
w-i-p

Commit:
w-i-p

Commit:
Establish a table.

Commit:
w-i-p

Commit:
w-i-p

gsps1 commented on cdap-docs/admin-manual/source/operations/logging.rst:
```diff
@@ -10,35 +10,106 @@ Logging and Monitoring
+multiple  programs distributed across the nodes of a cluster, the complete logs for the
+application may similarly be distributed.
 
+To address these issues, the CDAP log framework was designed:
```
not all the points below are directly related to the above issues, like pluggability is not related to this.

awholegunch replied:
Fixed.

gsps1 commented on cdap-docs/admin-manual/source/operations/logging.rst:
```diff
@@ -10,35 +10,106 @@ Logging and Monitoring
+  #. The Kafka topics are consumed by the Log Saver Service. Each instance of the service
+     has one or more logging pipelines. Each pipeline buffers log messages in-memory and
+     sorts them based on their timestamp. Custom pipelines can be configured by adding an
+     XML file in a prescribed location.
```
Log saver service is configured to read log messages for "logs.user-v2" kafka topic. 
Log saver instance can be scaled to process the kafka partitions in parallel, if needed. 
Log.saver reads the messages from kafka, buffers and sorts them in memory, before persisting to HDFS. These messages are written to files corresponding to the program the log messages belong to. 

Note: these files are configured to rotate based on time and size, they can be changed using these properties - log.pipeline.cdap.file.max.size.bytes and log.pipeline.cdap.file.max.lifetime.ms

The files are written with permission such that they are accessible only by the "cdap" user for security reasons.

Log.saver also emits metrics for each {info,debug,error} messages received. this helps users to monitor the amount of logs emitted by application,  application's programs, system services, etc by querying CDAP metrics system.

awholegunch replied:
Added

gsps1 commented on cdap-docs/admin-manual/source/operations/logging.rst:
```diff
@@ -10,35 +10,106 @@ Logging and Monitoring
+<https://logback.qos.ch/manual>`__, and consists of *logging pipelines* with *appenders*:
+
+  - A **logging pipeline** is a process that consumes log events from Kafka, invokes
+    an *appender* defined in its configuration, buffering, sorting, and then persisting the log messages.
```
appenders, there can be multiple.

awholegunch replied:
Fixed.

gsps1 commented on cdap-docs/admin-manual/source/operations/logging.rst:
```diff
@@ -10,35 +10,106 @@ Logging and Monitoring
+  - A **logging pipeline** is a process that consumes log events from Kafka, invokes
+    an *appender* defined in its configuration, buffering, sorting, and then persisting the log messages.
+
+  - An **appender** is a Java class, responsible for writing the log events to persistent storage and
```
hmm, appender is a "logback appender" is a better description i think ?

awholegunch replied:
I was trying to match the property name ``appender``.

awholegunch replied:
Adjusted.

gsps1 commented on cdap-docs/admin-manual/source/operations/logging.rst:
```diff
@@ -10,35 +10,106 @@ Logging and Monitoring
+- The ``logback-container.xml`` controls the program containers, both locally and to Kafka.
+- The ``cdap-site.xml`` controls the CDAP Log Pipeline for CDAP System Services, and the
+  reading, buffering, and memory allocated.
+- A ``custom-pipeline.xml``, if present, triggers the creation of a custom pipeline, with
```
there can be any number of `xml` files with any name, this description sound like we look for this name.

awholegunch replied:
Edited to remove that name.

gsps1 commented on cdap-docs/admin-manual/source/operations/logging.rst:
```diff
@@ -50,28 +121,234 @@ CDAP supports logging through standard `SLF4J (Simple Logging Facade for Java)
+
+- For Standalone CDAP: ``<cdap-sdk-home>/conf/logback.xml``
+- For Distributed CDAP: ``/opt/cdap/master/ext/logging/config``, as set by the property
+  ``log.process.pipeline.config.dir`` in the ``cdap-default.xml`` file
```
these are for adding new logging pipelines. not relevant to configuring logging.

gsps1 commented on cdap-docs/admin-manual/source/operations/logging.rst:
```diff
@@ -50,28 +121,234 @@ CDAP supports logging through standard `SLF4J (Simple Logging Facade for Java)
+These properties control the writing of logs to Kafka:
+
+- ``log.publish.num.partitions`` (default 10)
+- ``log.publish.partition.key``
```
missing description

awholegunch replied:
Fixed.

gsps1 reviewed (commented):
looked at logging.rst file. needs improvement. in general i think its better to talk about application logging at the begging, user would edit `logback-container.xml` more often than add new logback.xml for custom logging.

Commit:
edits per comments.

Commit:
w-i-p

Commit:
w-i-p

Commit:
Update index page

Commit:
w-i-p

Commit:
Adding details.

Commit:
Change illustration

Commit:
Delete file accidentally included.

anew commented on cdap-docs/admin-manual/source/index.rst:
```diff
@@ -115,9 +115,9 @@ installation and its security configuration.
-      Logging Facade for Java) APIs.
+      Covers **log location**, **logging messages**, the **system services and user
+      application logback configuration** and **CDAP support for logging** through standard
+      SLF4J (Simple Logging Facade for Java) APIs and Logback.
```
does it make sense to add links here?

awholegunch replied:
Yes; were you thinking internal links, such as such as to ``log location``, ``logging messages``, ``system services and user
application logback configuration``, and ``CDAP support for logging``?

awholegunch replied:
Added external links.

anew commented on cdap-docs/developers-manual/source/advanced/application-logback.rst:
```diff
@@ -8,19 +8,23 @@
+in the ``<cdap-sdk-home>/conf`` directory.
 
-However, you can specify a custom ``logback.xml`` for a CDAP application by packaging
+You can specify a custom ``logback.xml`` for a CDAP application by packaging
```
Is this correct? Or does the file name have to be logback-container.xml? (I don't know this... but it seems to make more sense)

awholegunch replied:
@gsps1 ? ?

gsps1 replied:
i checked with @yaojiefeng and @poornachandra, it has to be `logback.xml` for logback to pick it up.

awholegunch replied:
Filename must be `logback.xml`.

anew commented on cdap-docs/admin-manual/source/operations/logging.rst:
```diff
@@ -1,44 +1,96 @@
-Logs and Metrics
-================
-CDAP collects logs and metrics for all of its internal services and and user applications.
+CDAP collects logs and metrics for all of its internal services and on user applications.
```
remove 'on'?

awholegunch replied:
Fixed.

anew commented on cdap-docs/admin-manual/source/operations/logging.rst:
```diff
@@ -1,44 +1,96 @@
-and :ref:`Monitoring <http-restful-api-monitor>` HTTP RESTful APIs, the :ref:`Java Client
-<reference:java-client-api>`, and the :ref:`master-service-logging-configuration` for
-additional information.
+In Hadoop clusters, the programs running in containers generate individual log files as
```
inside their containers?

awholegunch replied:
Fixed.

anew commented on cdap-docs/admin-manual/source/operations/logging.rst:
```diff
@@ -1,44 +1,96 @@
+In Hadoop clusters, the programs running in containers generate individual log files as
+part of the container. As these files generally do not persist beyond the life of the
+container, they are volatile and not very useful for post-mortem diagnostics and
+troubleshooting, or for analyzing performance. As a single application can consist of
```
This second sentence seems disconnected and unrelated to the previous one. Would you want to say, "Also, as an application can consist..."

anew replied:
Perhaps better use the word program. An app can have multiple programs, and we collect for each program.

anew replied:
Is an additional benefit that we will make the logs available while the program is still running? What is the benefit over using "yarn logs -applicationId ..."?

awholegunch replied:
Fixed.

anew commented on cdap-docs/admin-manual/source/operations/logging.rst:
```diff
@@ -1,44 +1,96 @@
-Log Location
-============
-The location of logs depends on the mode of CDAP and the Hadoop distribution.
+- to centralize the location of logs, so that the individual logs of an application's programs can be
```
the logs of the individual containers of a program can be

awholegunch replied:
Fixed.

anew commented on cdap-docs/admin-manual/source/operations/logging.rst:
```diff
@@ -1,44 +1,96 @@
-  directories under ``/var/run/cloudera-scm-agent/process``.
+- to be extensible using :ref:`custom log pipelines <logging-monitoring-custom-logging-pipeline>`; and
 
+- to use a system that can be adapted for individual requirements at both the application
```
to allow fine-tuning the logging behavior at the level of an individual application as well as the entire cluster

awholegunch replied:
Fixed.

anew commented on cdap-docs/admin-manual/source/operations/logging.rst:
```diff
@@ -1,44 +1,96 @@
+
+  - YARN containers have their own local log files.
+  
+  - CDAP logs are published to Kafka.
```
This sounds as if YARN logs and CDAP logs are disjoint sets of logs. You may better say: 
- YARN writes the log messages emitted by containers to files inside the container
- In addition, CDAP programs publish these messages to Kafka

awholegunch replied:
Fixed.

anew commented on cdap-docs/admin-manual/source/operations/logging.rst:
```diff
@@ -1,44 +1,96 @@
+  - CDAP logs are published to Kafka.
+
+  - The Log Saver Service (*Log.saver*) is configured to read log messages from Kafka.
+    Log.saver reads the messages from Kafka, buffers and sorts them in-memory, before
```
in memory

awholegunch replied:
Fixed.

anew commented on cdap-docs/admin-manual/source/operations/logging.rst:
```diff
@@ -1,44 +1,96 @@
+  - CDAP logs are published to Kafka.
+
+  - The Log Saver Service (*Log.saver*) is configured to read log messages from Kafka.
+    Log.saver reads the messages from Kafka, buffers and sorts them in-memory, before
```
groups them by program or application, buffers and sorts them in memory, and finally persists them to files in HDFS.

awholegunch replied:
Fixed.

anew commented on cdap-docs/admin-manual/source/operations/logging.rst:
```diff
@@ -1,44 +1,96 @@
+
+  - The Log Saver Service (*Log.saver*) is configured to read log messages from Kafka.
+    Log.saver reads the messages from Kafka, buffers and sorts them in-memory, before
+    persisting them to HDFS. These messages are written to files corresponding to the
```
Each of these files corresponds to one program or application, depending on how the grouping is configured. (see reference)

awholegunch replied:
Fixed.

anew commented on cdap-docs/admin-manual/source/operations/logging.rst:
```diff
@@ -1,44 +1,96 @@
+    persisting them to HDFS. These messages are written to files corresponding to the
+    programs the log messages belong to.
+
+  - Metrics are created based on the results of the pipeline, and passed to the Metrics Service.
```
- In addition to persisting logs to files, the Log.saver also emits metrics about the number of log messages emitted by each program.

awholegunch replied:
Fixed.

anew commented on cdap-docs/admin-manual/source/operations/logging.rst:
```diff
@@ -1,44 +1,96 @@
+  - For security, the files written out to persistent storage in HDFS have permissions set
+    such that they are accessible only by the ``cdap`` user.
+
+  - Log.saver also emits metrics for each *info, debug,* or *error* message received. This
```
Is this the same as the metrics mentioned above? In that case, the bullet above is redundant.

awholegunch replied:
Fixed.

anew commented on cdap-docs/admin-manual/source/operations/logging.rst:
```diff
@@ -1,44 +1,96 @@
+CDAP supports application logging through the standard `SLF4J (Simple Logging Facade for Java)
+<http://www.slf4j.org/manual.html>`__ APIs.
+
+For instance, in a flowlet you can write::
```
I think we should change this to debug instead of info (or perhaps even trace). We do not want to encourage excessive info logging.

awholegunch replied:
Fixed.

anew commented on cdap-docs/admin-manual/source/operations/logging.rst:
```diff
@@ -50,28 +102,409 @@ CDAP supports logging through standard `SLF4J (Simple Logging Facade for Java)
+file, packaged with the CDAP distribution. This "logback" does log rotation once a day at
+midnight and expires logs older than 14 days. Changes can be made to
+``logback-container.xml``; afterwards, the ``cdap-master`` process will need to be
+restarted for the modified logback file to take effect.
```
Is that really true? Doesn't master read it every time it starts a program?

awholegunch replied:
Changed to "afterwards, applications or programs needs to be restarted…"

anew commented on cdap-docs/admin-manual/source/operations/logging.rst:
```diff
@@ -50,28 +102,409 @@ CDAP supports logging through standard `SLF4J (Simple Logging Facade for Java)
+file, packaged with the CDAP distribution. This "logback" does log rotation once a day at
+midnight and expires logs older than 14 days. Changes can be made to
+``logback-container.xml``; afterwards, the ``cdap-master`` process will need to be
+restarted for the modified logback file to take effect.
```
It should also be mentioned that changing logback-container.xml will only affect programs that are started after the change; existing running programs will not be affected.

awholegunch replied:
Fixed.

anew commented on cdap-docs/admin-manual/source/operations/logging.rst:
```diff
@@ -50,28 +102,409 @@ CDAP supports logging through standard `SLF4J (Simple Logging Facade for Java)
+for a particular run of a program. Once changed, they can be reset back to what they
+started with by using the :ref:`reset endpoint <http-restful-api-logging-resetting>`.
+
+To change the log level before an application starts, you can add the logger name as the
```
To configure

awholegunch replied:
Fixed.

anew commented on cdap-docs/admin-manual/source/operations/logging.rst:
```diff
@@ -50,28 +102,409 @@ CDAP supports logging through standard `SLF4J (Simple Logging Facade for Java)
+key and log level as the value in the preferences. The logger name should be prefixed with
+``system.log.level``. 
+
+For example, if you want to change the log level of the ``HelloWorld`` class of the
```
you want to set

awholegunch replied:
Fixed.

anew commented on cdap-docs/admin-manual/source/operations/logging.rst:
```diff
@@ -50,28 +102,409 @@ CDAP supports logging through standard `SLF4J (Simple Logging Facade for Java)
+
+To change the log level of a program dynamically (for a worker, flow, or service which are
+running), see :ref:`Logging HTTP RESTful API
+<http-restful-api-logging-changing-program-log-levels>`. You can change the log level at
```
Repetition here?

awholegunch replied:
Fixed.

anew commented on cdap-docs/admin-manual/source/operations/logging.rst:
```diff
@@ -50,28 +102,409 @@ CDAP supports logging through standard `SLF4J (Simple Logging Facade for Java)
+To change the log level of a program dynamically (for a worker, flow, or service which are
+running), see :ref:`Logging HTTP RESTful API
+<http-restful-api-logging-changing-program-log-levels>`. You can change the log level at
+run level of a worker, flow, or service.
```
you mean, at run time?

awholegunch replied:
Fixed.

anew commented on cdap-docs/admin-manual/source/operations/logging.rst:
```diff
@@ -50,28 +102,409 @@ CDAP supports logging through standard `SLF4J (Simple Logging Facade for Java)
+<http-restful-api-logging-changing-program-log-levels>`. You can change the log level at
+run level of a worker, flow, or service.
+
+**Note:** The log levels can only be changed for programs that are running under
```
you mean, before a program starts? Or once it's running? Or both?

I find it a little confusing to speak of "change" in both cases.

awholegunch replied:
Fixed (I think!)

anew commented on cdap-docs/admin-manual/source/operations/logging.rst:
```diff
@@ -50,28 +102,409 @@ CDAP supports logging through standard `SLF4J (Simple Logging Facade for Java)
+
+Changing Program Log Levels
+---------------------------
+The CDAP :ref:`Logging HTTP RESTful API
```
Is this for running programs, or for before a program starts?

awholegunch replied:
Turns out this will work for a _running_ program. I will try to clear this up.

anew commented on cdap-docs/admin-manual/source/operations/logging.rst:
```diff
@@ -50,28 +102,409 @@ CDAP supports logging through standard `SLF4J (Simple Logging Facade for Java)
+- Using the :ref:`restful-api`: the :ref:`Logging HTTP RESTful API
+  <http-restful-api-logging>` details :ref:`downloading the logs
+  <http-restful-api-logging-downloading-system-logs>` emitted by a system service.
+- Log messages of system services can be viewed in the :ref:`CDAP UI <cdap-ui>`.
```
Can we be more specific? How do I navigate to these logs?

awholegunch replied:
Added images and details. Also added for program logs.

awholegunch replied:
Fixed.

anew commented on cdap-docs/admin-manual/source/operations/logging.rst:
```diff
@@ -50,28 +102,409 @@ CDAP supports logging through standard `SLF4J (Simple Logging Facade for Java)
+
+Configuring System Service Logs
+-------------------------------
+- To configure logging by CDAP system services that run in YARN containers, see the
```
This appears like a repetition of lines 178ff

awholegunch replied:
Fixed.

anew commented on cdap-docs/admin-manual/source/operations/logging.rst:
```diff
@@ -50,28 +102,409 @@ CDAP supports logging through standard `SLF4J (Simple Logging Facade for Java)
-for logging configuration. This logback does log rotation once a day at midnight and expires logs older than
-14 days. Changes can be made to ``logback-container.xml``, and the ``cdap-master`` process will need to be restarted
-for the modified logback to take effect.
+**Custom logging pipelines** are configured by a combination of the settings in the 
```
Did you introduce/motivate the term "custom log pipeline" previously? It seems to fall out of the sky here.

awholegunch replied:
Yes, it has. Fixed by adding narrative previous to this.

anew commented on cdap-docs/admin-manual/source/operations/logging.rst:
```diff
@@ -50,28 +102,409 @@ CDAP supports logging through standard `SLF4J (Simple Logging Facade for Java)
+**Custom logging pipelines** are configured by a combination of the settings in the 
+:ref:`cdap-site.xml <appendix-cdap-site.xml>` file and a "logback" file used
+to specify the custom pipeline. The XML file is placed in the
+``log.process.pipeline.config.dir``, a local directory on the CDAP Master node that is scanned
```
This is a property in cdap-site.xml, right? Can you add what the default is?

awholegunch replied:
Added.

anew commented on cdap-docs/admin-manual/source/operations/logging.rst:
```diff
@@ -50,28 +102,409 @@ CDAP supports logging through standard `SLF4J (Simple Logging Facade for Java)
+
+  - Logs are collected from individual programs running in YARN containers.
+
+  - YARN containers have their own local log files.
```
change this section according to the changes in further up (where you describe the simple case)

awholegunch replied:
Updated.

anew commented on cdap-docs/admin-manual/source/operations/logging.rst:
```diff
@@ -50,28 +102,409 @@ CDAP supports logging through standard `SLF4J (Simple Logging Facade for Java)
+    :ref:`logging-monitoring-logging-pipeline-configuration`.
+
+  - Custom pipelines can be configured by adding an XML file in a prescribed location.
+    Each pipeline buffers log messages in-memory and sorts them based on their timestamp. 
```
in memory

anew replied:
Perhaps structure this as follows:
- Log.saver, by default, runs only the CDAP log pipeline. This performs:
  - read, group, sort, persist to files
  - emit metrics about each program/log level
- In addition, custom log pipelines can be configured...

awholegunch replied:
Fixed.

anew commented on cdap-docs/admin-manual/source/operations/logging.rst:
```diff
@@ -50,28 +102,409 @@ CDAP supports logging through standard `SLF4J (Simple Logging Facade for Java)
+<https://logback.qos.ch/manual>`__. Logging is configured using instances of Logback's
+"logback" file, consisting of *logging pipelines* with *logging appenders*:
+
+  - A **logging pipeline** is a process that consumes log events from Kafka, invokes
```
Is it a "log pipeline" or a "logging pipeline"? Better make it the same everywhere

awholegunch replied:
Changed to "log pipeline" and "log appender".

anew commented on cdap-docs/admin-manual/source/operations/logging.rst:
```diff
@@ -50,28 +102,409 @@ CDAP supports logging through standard `SLF4J (Simple Logging Facade for Java)
+    persisting the log messages.
+
+  - A **logback appender** (or *appender*) is a Java class, responsible for writing the
+    log events to persistent storage and maintaining metadata about that storage.
```
not necessarily persisted. What it really does is consume and process messages, which typically includes/implies persisting them; however it could also collect metrics, or emit alerts when it finds certain messages.

awholegunch replied:
Fixed.

anew commented on cdap-docs/admin-manual/source/operations/logging.rst:
```diff
@@ -50,28 +102,409 @@ CDAP supports logging through standard `SLF4J (Simple Logging Facade for Java)
+- The ``cdap-site.xml`` controls the CDAP Log Pipeline for CDAP System Services, and the
+  reading, buffering, and memory allocated.
+- Custom "logback" files, if present in a specified directory, trigger the creation of
+  custom logging pipelines, with a pipeline created for each custom file.
```
This whole paragraph repeats information from further above. is that on purpose?

awholegunch replied:
Deleted!

anew commented on cdap-docs/admin-manual/source/operations/logging.rst:
```diff
@@ -50,28 +102,409 @@ CDAP supports logging through standard `SLF4J (Simple Logging Facade for Java)
+
+Custom Logging Pipeline
+=======================
+For a custom logging pipeline, you would create and configure a Logback XML file,
```
remove "you would"

awholegunch replied:
Fixed.

anew commented on cdap-docs/admin-manual/source/operations/logging.rst:
```diff
@@ -50,28 +102,409 @@ CDAP supports logging through standard `SLF4J (Simple Logging Facade for Java)
+<custom-log-pipeline-logging-configuration>`.
+
+For every XML file in the ``log.process.pipeline.config.dir`` directory, a separate logging pipeline is created, providing
+isolation from other logging pipelines. As they are separate Kafka consumers and
```
Does this isolation mean that they run as separate processes? If one crashes, the other one survives? Or what exactly is this isolation?

anew commented on cdap-docs/admin-manual/source/operations/logging.rst:
```diff
@@ -50,28 +102,409 @@ CDAP supports logging through standard `SLF4J (Simple Logging Facade for Java)
+
+.. _logging-monitoring-custom-logging-configuration:
+
+Custom Logging Pipeline Configuration Pipelines
```
Strange title....

awholegunch replied:
Fixed.

anew commented on cdap-docs/admin-manual/source/operations/logging.rst:
```diff
@@ -50,28 +102,409 @@ CDAP supports logging through standard `SLF4J (Simple Logging Facade for Java)
+
+Custom Logging Pipeline Configuration Pipelines
+-----------------------------------------------
+CDAP looks for Logback "logback" files located in a directory as set by the property
```
do you mean it looks for xml files?

awholegunch replied:
Fixed.

anew commented on cdap-docs/admin-manual/source/operations/logging.rst:
```diff
@@ -50,28 +102,409 @@ CDAP supports logging through standard `SLF4J (Simple Logging Facade for Java)
+.. highlight:: xml
+
+Here is an example "logback" file, using two appenders (``STDOUT`` and
+``rollingAppender``). It would be saved with the file extension ``.xml`` to a location as
```
"would be saved" sounds strange. Better: This file must be located in the location ... with a file extension of ".xml"

awholegunch replied:
Fixed.

anew commented on cdap-docs/admin-manual/source/operations/logging.rst:
```diff
@@ -50,28 +102,409 @@ CDAP supports logging through standard `SLF4J (Simple Logging Facade for Java)
+As the CDAP LogFramework uses Logback's Appender API, your custom appender needs to
+implement the same Appender interface. Access to CDAP's system components (such as
+datasets, metrics, ``LocationFactory``) are made available to the ``AppenderContext``, an
+extension of Logback's ``LoggerContext``.
```
and how do I get access to that AppenderContext? Code example?

awholegunch replied:
Added.

anew reviewed (changes requested):
Overall this is very good and comprehensive.

I have quite a few comments. Especially there appears to be a lot of repeated content, which is a little confusing to readers. 

Also, because I can't make a comment on an image. Regarding the logging-framework.png, I am not sure that the CDAP Master Service writes its log messages to Kafka. I think only containers write to Kafka. Can you verify that?

awholegunch commented on cdap-docs/admin-manual/source/operations/logging.rst:
```diff
@@ -50,28 +102,409 @@ CDAP supports logging through standard `SLF4J (Simple Logging Facade for Java)
+    ``logs.user-v2`` Kafka topic. The number of log saver instances can be scaled to process
+    the Kafka partitions in parallel, if needed. Log.saver reads the messages from Kafka,
+    buffers and sorts them in-memory, before persisting them to HDFS. These messages are
+    written to files corresponding to the programs the log messages belong to.
```
log.publish.partition.key

awholegunch replied:
Added.

awholegunch commented on cdap-docs/admin-manual/source/operations/index.rst:
```diff
@@ -27,8 +27,9 @@ Operations
-  **CDAP support for logging** through standard SLF4J (Simple Logging Facade for Java) APIs.
+  **log location**, **logging messages**, the **system services and user application logback
+  configuration** and **CDAP support for logging** through standard SLF4J (Simple Logging
+  Facade for Java) APIs and Logback.
```
ibid.

gsps1 commented:
the cdap services that run on edge node uses `logback.xml` its not accurate to denote that as CDAP master services alone, as router, auth-server, etc also use the same config.

About @anew comment, CDAP master services like app.fabric etc that run on edge node still write to kafka for log aggregation. but don't think we need to get into that level of detail, we can simplify by calling that component as CDAP services (running on edge node (or) better term) and don't link that to kafka.

Commit:
Edits per comments.

awholegunch commented:
Quick Build: http://builds.cask.co/browse/CDAP-DQB303-11

Page of interest: **Logging and Monitoring:** http://builds.cask.co/artifact/CDAP-DQB303/shared/build-11/Docs-HTML/4.1.0/en/admin-manual/operations/logging.html

anew commented:
In the simple logging example drawing: The logback-container.xml controls all logging in the container (logging to files, as well as - if desired - additional log appenders). It does not control the logging to Kafka. It does control the log level for different loggers (a logger in this context means a Java class or package). 

In this light, and to keep it simple, can we just say, "controls how a program container emits logs"?

anew commented on cdap-docs/admin-manual/source/operations/logging.rst:
```diff
@@ -1,77 +1,573 @@
+"logback" file, consisting of *log pipelines* with *log appenders*:
+
+  - A **log pipeline** is a process that consumes log events from Kafka, invokes
+    *logback appenders* defined in its configuration, buffering, sorting, and then
```
first: buffers, groups by app/program and sorts, then invokes log appenders.

awholegunch replied:
Fixed.

anew commented on cdap-docs/admin-manual/source/operations/logging.rst:
```diff
@@ -1,77 +1,573 @@
+    *logback appenders* defined in its configuration, buffering, sorting, and then
+    persisting the log messages.
+
+  - A **logback appender** (or *appender*) is a Java class, responsible for consuming and
```
should this be "log appender"?

awholegunch replied:
Fixed.

anew commented on cdap-docs/admin-manual/source/operations/logging.rst:
```diff
@@ -1,77 +1,573 @@
+
+  - A **logback appender** (or *appender*) is a Java class, responsible for consuming and
+    processing messages; typically, this includes persisting the log events. It can also
+    collect metrics, maintain metadata about the storage, or emit alerts when it finds
```
, for example,

awholegunch replied:
Fixed.

anew commented on cdap-docs/admin-manual/source/operations/logging.rst:
```diff
@@ -1,77 +1,573 @@
   }
 
-The log messages emitted by your application code can be viewed in different ways.
+This will emit "debug" level messages from a flowlet when it is processing an event.
```
from the flowlet

awholegunch replied:
Fixed.

anew commented on cdap-docs/admin-manual/source/operations/logging.rst:
```diff
@@ -1,77 +1,573 @@
+      :align: center
+      :class: bordered-image
+
+      **CDAP UI:** *Log Viewer*, showing *PurchaseFlow* log events, *INFO* level
```
the image here appears large with a lot of gray space. Could you take the screenshot from a smaller window?

awholegunch replied:
Fixed.

anew commented on cdap-docs/admin-manual/source/operations/logging.rst:
```diff
@@ -1,77 +1,573 @@
+  ``${local.data.dir}/logs``) is the root location for collecting program logs when in
+  Standalone CDAP.
+
+- For **Distributed CDAP:** the property ``log.base.dir`` (default ``/logs/avro``) is the
```
seriously? we use different properties? Are you sure?

awholegunch replied:
That's what the CDAP Defaults say...

gsps1 replied:
we don't use `log.base.dir` in distributed anymore. its `<cdap-dir>/logs` cdap-dir is based on cdap/hdfs namespace config.

awholegunch replied:
Fixed.

anew commented on cdap-docs/admin-manual/source/operations/logging.rst:
```diff
@@ -1,77 +1,573 @@
+be affected.
+
+- For **Distributed CDAP:** the ``logback-container.xml`` file is located in ``/etc/cdap/conf``.
+- For **Standalone CDAP:** as containers are replaced with Java threads, a single
```
perhaps a little less technical: As the entire Standalone CDAP runs in one JVM, a single...

awholegunch replied:
Fixed.

anew commented on cdap-docs/admin-manual/source/operations/logging.rst:
```diff
@@ -1,77 +1,573 @@
+  and CDAP system service logging.
+  
+You can also use a custom "logback" file with your application, as described in the
+Developers' Manual section :ref:`application-logback`.
```
and does that work in both Standalone and Distributed?

awholegunch replied:
According to @yaojiefeng , yes.

anew commented on cdap-docs/admin-manual/source/operations/logging.rst:
```diff
@@ -1,77 +1,573 @@
+only affect programs that are started after the change; existing running programs will not
+be affected.
+
+- For **Distributed CDAP:** the ``logback-container.xml`` file is located in ``/etc/cdap/conf``.
```
Keep the order consistent? First standalone, then distributed (as in the previous paragraph)

awholegunch replied:
Fixed.

anew commented on cdap-docs/admin-manual/source/operations/logging.rst:
```diff
@@ -1,77 +1,573 @@
+<http-restful-api-logging-resetting>`.
+
+- To configure the log level before an application starts, you can add the logger name as the
+  key and log level as the value in the preferences. The logger name should be prefixed with
```
add a link to the preferences REST API and UI?

awholegunch replied:
Added.

anew commented on cdap-docs/admin-manual/source/operations/logging.rst:
```diff
@@ -1,77 +1,573 @@
+  <http-restful-api-logging-downloading-system-logs>` emitted by a system service.
+- Log messages of system services can be viewed in the :ref:`CDAP UI <cdap-ui>`.
+
+- Log messages of a program can be viewed in the :ref:`CDAP UI <cdap-ui>`. In the
```
hmmm. these don't seem to be programs but Hadoop infrastructure components.

awholegunch replied:
Fixed. Bad copy and paste. That sentence was suppose to be replaced by line 219.

anew commented on cdap-docs/admin-manual/source/operations/logging.rst:
```diff
@@ -1,77 +1,573 @@
+  configured by the same ``logback-container.xml`` that :ref:`configures user application
+  program logging <configuring-user-application-program-logging>`.
+
+- CDAP system services that run on cluster edge nodes, such as CDAP Master or Router, are
```
this is confusing... it requires that the user knows where a service is running (container or edge node). Wouldn't it make a lot more sense to say "logback.xml" configures CDAP, whereas "logback-container.xml" or a "logback.xml" bundled with the app configures app containers? Perhaps even better to name it logback-app.xml, and the same bundled in the app. 

Would you agree? In that case we should file a Jira so that this can be improved.

awholegunch replied:
I agree this is confusing. It does require users to know which services are running in YARN containers, and which aren't. I would recommend a Jira to clean this up, though if that should be one file or two different ones I don't know.

I don't think saying *"logback.xml" configures CDAP* helps, as what does "configure CDAP" mean? But can't we list which services are in each location? Even if we only list one set, and say that the other set is everything else..

anew replied:
That sounds right. However, I am not even sure what that list is, and how will we maintain it over time when new services are added? I think you can leave this as is for now. I have created https://issues.cask.co/browse/CDAP-9138 for the improvement.

anew commented on cdap-docs/admin-manual/source/operations/logging.rst:
```diff
@@ -1,77 +1,573 @@
 
-.. _master-service-logging-configuration:
+*Note:* By default, ``log.publish.partition.key`` is set to ``program``, which means that
+all logs for the same program (and program type) go to the same partition.
```
remove "(and program type)". Each program has one name and one type, and therefore this is redundant.

awholegunch replied:
Fixed.

anew commented on cdap-docs/admin-manual/source/operations/logging.rst:
```diff
@@ -1,77 +1,573 @@
+
+  - YARN writes the log messages emitted by containers to files inside the container.
+
+  - CDAP System Services run (depending on the service) either on cluster edge nodes or in
```
flip 3rd and 4th bullet. The "these messages" really refers to the same log messages that Yarn writes.

awholegunch replied:
Fixed.

anew commented on cdap-docs/admin-manual/source/operations/logging.rst:
```diff
@@ -1,77 +1,573 @@
+    :ref:`logging-monitoring-log-pipeline-configuration`.
+
+  - In addition, custom log pipelines can be configured by adding an XML file in a
+    prescribed location. Each pipeline buffers log messages in memory and sorts them based
```
what is that location? link?

anew replied:
ok , I see this is mentioned in the next section.

anew commented on cdap-docs/admin-manual/source/operations/logging.rst:
```diff
@@ -1,77 +1,573 @@
+    prescribed location. Each pipeline buffers log messages in memory and sorts them based
+    on their timestamp. 
+   
+  - In addition to persisting logs to files, the Log.saver also emits metrics about the
```
Do we document anywhere what the names of these metrics are? If so, a link would help here.

anew replied:
Are you planning to add these metrics or a link?

anew commented on cdap-docs/admin-manual/source/operations/logging.rst:
```diff
@@ -1,77 +1,573 @@
+file at the path specified in the ``cdap-site.xml`` file by the property
+``log.process.pipeline.config.dir`` of the ``cdap-site.xml`` file.
+
+Each custom pipeline requires a unique name, which is used for persisting the data and the
```
a unique name. Properties controlling...

awholegunch replied:
Fixed.

anew commented on cdap-docs/admin-manual/source/operations/logging.rst:
```diff
@@ -1,77 +1,573 @@
+<logging-monitoring-cdap-site-xml-configuration>`.
+
+For every XML file in the ``log.process.pipeline.config.dir`` directory, a separate log
+pipeline is created, providing isolation from other log pipelines. As they are separate
```
I would prefer a colon: ... isolation from other pipelines: As they are...

awholegunch replied:
Changed.

anew commented on cdap-docs/admin-manual/source/operations/logging.rst:
```diff
@@ -1,77 +1,573 @@
+
+Configuring Custom Log Pipelines
+--------------------------------
+CDAP looks for Logback XML files located in a directory as set by the property
```
is it a "Logback XML file" or a "'logback' file"? Perhaps you can use the same term everywhere. (I don't have preference for either one).

awholegunch replied:
Fixed. Changed other instances to "logback", leaving only one instance of "Logback XML" as it was appropriate.

anew commented on cdap-docs/admin-manual/source/operations/logging.rst:
```diff
@@ -1,77 +1,573 @@
+and implement your own custom appender. See the `Logback documentation
+<https://logback.qos.ch/manual/appenders.html>`__ for information on this.
+
+You can use any existing `logback <https://logback.qos.ch/manual/appenders.html>`__
```
switch this with the previous paragraph

awholegunch replied:
Fixed.

anew reviewed (changes requested):
This is really looking great now, and the structure is clear and easy to follow. I do still have a few comments...

yaojiefeng commented on cdap-docs/admin-manual/source/operations/logging.rst:
```diff
@@ -1,77 +1,573 @@
+---------------------------
+The CDAP :ref:`Logging HTTP RESTful API
+<http-restful-api-logging-changing-program-log-levels>` can be used to set the log levels
+of a program, either before a particular run or while it is running. Once changed, they
```
To change log levels of a running program, the program type has to be one of service, worker, flow. Other program types are currently not supported. Their log levels can only be changed using preferences before the program starts.

awholegunch replied:
Added.

yaojiefeng commented on cdap-docs/reference-manual/source/http-restful-api/logging.rst:
```diff
@@ -163,6 +163,8 @@ A filter string can be supplied as an optional parameter. It will filter the ret
+
 Changing Program Log Levels
 ===========================
 Log levels can be set for a particular run of a program. Once changed, they can be reset back to what
```
This is only supported for program type flow, service or worker. The corresponding programType at Line 210 should be updated accordingly

awholegunch replied:
Fixed. Changed lines 210 and 265 accordingly.

Commit:
Changes per updates.

awholegunch commented:
Started a new build: http://builds.cask.co/browse/CDAP-DQB303-12

yaojiefeng commented:
Change program log level parts LGTM

anew commented:
My only remaining (un-addressed) comment is that we should also document what metrics are actually emitted by the log saver. Wouldn't that be useful? If you don't know, fine to open a Jira to do it in a different PR. 

Other than that: LGTM

Commit:
Added metrics tables. Added notes about copying. Moved figure so that it
appears before the commentary. Moved not about security to be before the
tables so that it doesn't get lost.

awholegunch commented:
Started a new build: http://builds.cask.co/browse/CDAP-DQB303-13

Page of interest (and section): http://builds.cask.co/artifact/CDAP-DQB303/shared/build-13/Docs-HTML/4.1.0/en/admin-manual/operations/logging.html#logging-framework

anew commented on cdap-docs/admin-manual/source/operations/logging.rst:
```diff
@@ -410,6 +371,79 @@ programs of an application and system services with the logging framework:
+  prescribed location. Each pipeline buffers log messages in memory and sorts them based
+  on their timestamp. 
+ 
+- In addition to persisting logs to files, the *Log.saver* also emits metrics about the
```
in the light of recent discussions, our documentation should use english words to describe things. Can you call this the "Log Saver" everywher instead of "Log.saver"?

awholegunch replied:
I've removed all but one, as you do need to know about that name (_log.saver_) to access the service using the RESTful APIs.

anew commented on cdap-docs/admin-manual/source/operations/logging.rst:
```diff
@@ -410,6 +371,79 @@ programs of an application and system services with the logging framework:
+
+     * - **Service Metric**
+       - **Description**
+     * - ``system.requests.count``
```
In this context, it only seems to make sense to talk about the last two in this table. The other are irrelevant to logging? Or aren't they?

awholegunch replied:
The last two are probably the only ones relevant to users, even if the others are valid.

anew commented on cdap-docs/admin-manual/source/operations/logging.rst:
```diff
@@ -410,6 +371,79 @@ programs of an application and system services with the logging framework:
+       - Number of ``server-error`` or ``client-error`` responses sent
+     * - **Application Logging Metric**
+       - 
+     * - ``system.app.log.{error, info, warn}``
```
Is this true? We use different metrics names for the logs emitted by system services than for the logs emitted by apps? I find that hard to believe. Is this verified?

anew replied:
@chtyim do you know what would be the reason for this?

anew commented on cdap-docs/admin-manual/source/operations/logging.rst:
```diff
@@ -410,6 +371,79 @@ programs of an application and system services with the logging framework:
+     * - **Application Logging Metric**
+       - 
+     * - ``system.app.log.{error, info, warn}``
+       - Number of ``error``, ``info``, or ``warn`` log messages logged by the application
```
I think this should be "by an application", or "by applications"

anew replied:
Reason is that I can retrieved these metrics aggregated across all apps. In that case, "the application" is inaccurate because it is representing all applications. But when I look at the metrics for a single app, "all applications" does not make sense. So it is better to be a little indefinitive.

awholegunch replied:
Changed to be both:
```
- Number of ``error``, ``info``, or ``warn`` log messages logged 
  by an application or applications

anew commented on cdap-docs/admin-manual/source/operations/logging.rst:
```diff
@@ -410,6 +371,79 @@ programs of an application and system services with the logging framework:
+     * - **System Services Logging Metric**
+       - 
+     * - ``system.services.log.{error, info, warn}``
+       - Number of ``error``, ``info``, or ``warn`` log messages logged by the system services
```
and this should be "by a system service", or "by system services"

anew replied:
The reason is that they appear in different logging contexts. When I look at a single system service's logs, these metrics represent that particular service (not all the system services). Whereas if I look at these metrics at the global level, they represent all system services.

awholegunch replied:
Fixed, and also on the Metrics RESTful API page.

anew reviewed (changes requested):
I think we also need to rewrite all the descriptions of the properties that "control [both] the CDAP log pipeline[ and custom log pipelines]" tables. They currently describe implementation, not necessarily understandable by a user.

Commit:
Edits per comments. Fixed typo ("timeseries") in Metrics API.

awholegunch commented:
Started a build: http://builds.cask.co/browse/CDAP-DQB303-14

Page of interest: http://builds.cask.co/artifact/CDAP-DQB303/shared/build-14/Docs-HTML/4.1.0/en/admin-manual/operations/logging.html

Commit:
Edit per comments.

awholegunch commented:
Started a build: http://builds.cask.co/browse/CDAP-DQB303-15

Page of interest: http://builds.cask.co/artifact/CDAP-DQB303/shared/build-15/Docs-HTML/4.1.0/en/admin-manual/operations/logging.html

anew commented:
You use *Log saver* as if it were a proper name. Can't we just always call it "the log saver"? Just like we call the router "the router", and the master "the master".

awholegunch commented:
I have tried to call those "CDAP Master" and "CDAP Router", to distinguish them.  (From, say, other routers...) And in that same vein, calling it _Log Saver_.

anew commented:
But do you italicize (that's a funny word, but cursify sounds equally funny) the CDAP Master?

Commit:
Make only the first use of "Log Saver" in title case.

awholegunch commented:
I've gone through and removed the italicization of *Log Saver*, and made only the first usage in title case. You're correct on two points: italicize is an odd word, but it was incorrect and inconsistent to make it in italics.

anew commented:
LGTM and let's open a new PR for improving the descriptions of the logging configuration properties.

awholegunch commented:
Opened https://issues.cask.co/browse/CDAP-9184 for further work required.
